Add plugin: Co-Intelligence AI

mikethicke opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/Epistemic-Technology/co-intelligence

## Release Checklist
- [ ] I have tested the plugin on
  - [ ]  Windows
  - [x]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

mikethicke commented:
/skip The instances of assigning styles via Javascript are necessary for dynamically positioning a menu relative to a button. Other required changes have been addressed.

joethei commented:
["main": "main.js",](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/manifest.json#L11-L12)
These are not valid manifest entries, remove them.

[.workspace-tab-container {](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/src/styles.css#L481-L483)
Please don't overwrite core styling in a plugin, add another class/data attribute so it only applies to your plugin.

[const currentNoteContent = await app.vault.cachedRead(note);](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/src/utils/notes.ts#L260)
Use `.read` when you want to modify the file, ´.cachedRead` might be slightly out of date.

[function getAllTags(cache: CachedMetadata | null): string[] {](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/src/utils/notes.ts#L422-L441)
Use the `getAllTags` function from the Obsidian API instead.

[export function debounce<TArgs extends readonly unknown[]>(](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/src/utils/debounce.ts#L9-L28)
The Obsidian API already includes a debounce function.

[plugin,](https://github.com/Epistemic-Technology/co-intelligence/blob/9ddf244090643df85898760f64d4376fbf872892/src/components/MarkdownView.tsx#L44)
This should not be the plugin instance, the component passed to this function should be as short lived as possible, otherwise this will lead to memory leaks.

mikethicke commented:
Thanks for this. I wasn't aware of the Menu API---I'll give that a try.

mikethicke commented:
- removed main.js and main.css from manifest.json
- removed styling from Obsidian class and added to plugin class
- used read over cachedRead when modifying file
- replaced redundant getAllTags function with Obsidian version
- replaced redundant debounce function with Obsidian version
- use local instance of MarkDownRenderChild as component passed to renderMarkdown rather than plugin

Regarding the menu positioning, I investigated the Menu API, and unless I'm missing something (very possible!), I don't think it's suitable for my purposes. I have a screenshot below of the menu in question. It's currently implemented as a SolidJS component.

<img width="1120" height="689" alt="Screenshot 2025-08-20 at 3 29 17 PM" src="https://github.com/user-attachments/assets/011d53c8-234c-4c19-aa19-cf85cc079a3a" />

FWIW I do have a css-only implementation in this PR: https://github.com/Epistemic-Technology/co-intelligence/pull/2 , but the keyboard navigation in this version is really poor. It's more accessible and user friendly in my opinion to use the current method.

There is a PR here showing the changes I made: https://github.com/Epistemic-Technology/co-intelligence/pull/4/files

(It's cluttered with test files that can be ignored.)

Zachatoo commented:
[Copyright (c) 2025 Epistemic Technology](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/LICENSE#L3)
Please update this to the current year.

[.sr-only {](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/src/styles.css#L469)
Please update your classes to be more specific to your plugin to avoid conflicts with other plugins.

[new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/src/settings.ts#L201-L209), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/src/settings.ts#L250-L259), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/src/settings.ts#L261-L284)
Use [AbstractInputSuggest](https://docs.obsidian.md/Reference/TypeScript+API/AbstractInputSuggest) when letting a user select a file/folder to add type-ahead support.

---

[const renderChild = new MarkdownRenderChild(containerRef);](https://github.com/Epistemic-Technology/co-intelligence/blob/488a2eae86f99643e0bf47d2ba2ee54c3059db98/src/components/MarkdownView.tsx#L39)
This `renderChild` component is never being unloaded, this will result in a memory leak. It should be unloaded when it is no longer in use.

Zachatoo commented:
[new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L194-L202), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L247-L257), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L259-L284)
Use [AbstractInputSuggest](https://docs.obsidian.md/Reference/TypeScript+API/AbstractInputSuggest) when letting a user select a file/folder to add type-ahead support.

[new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L146-L156), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L158-L168), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L170-L180), [new Setting(containerEl)](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/settings.ts#L182-L192), ["minAppVersion": "1.8.0",](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/manifest.json#L5)
Please use [SecretStorage](https://docs.obsidian.md/plugins/guides/secret-storage) to securely store and manage sensitive data (requires `minAppVersion: "1.11.4"`)

[await app.vault.modify(note, newNoteContent);](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/utils/notes.ts#L271), [const currentNoteContent = await app.vault.read(note);](https://github.com/Epistemic-Technology/co-intelligence/blob/aba1ad821c81a067cc408891267611db214d9939/src/utils/notes.ts#L254)
Prefer to use [`Vault.process`](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+%60Vault.process%60+instead+of+%60Vault.modify%60+to+modify+a+file+in+the+background) to modify a file in the background.

mikethicke commented:
@Zachatoo -- Thank you for your feedback. I have made the requested changes:

- SecretStorage for API keys: API keys now stored in app.secretStorage instead of the unencrypted plugin data file. Includes a one-shot migration for existing users. Bumps minAppVersion to 1.11.4.

- AbstractInputSuggest for path settings: Folder and file path settings now use autocomplete instead of plain text inputs.

- Vault.process for note serialization: serializeCoiNote uses Vault.process for atomic read/write instead of separate read-then-modify calls.

See: https://github.com/Epistemic-Technology/co-intelligence/pull/15
